### PR TITLE
Identity / Add call to Discussion in PublicProfileController

### PR DESCRIPTION
**Pull request by vlbee**
## What does this change?

- Adds DiscussionApiService to Frontend/Identity
- Adds call to Discussion API in Identity's PublicProfileController to check if we should display a public profile for a given user or return a 404.

## Implementation Notes

- Adds a lightweight service to frontend/identity to make serverside GET requests to DiscussionApi (DAPI)
- Handles DAPI responses including 404 USER_NOT_FOUND - a Guardian User is only be created in Discussion when they either comment or recommend a comment for the first time
    
## Tested

- [x] Locally
- [x] On CODE
   - [x] Not in Discussion returns 404
  - [x] In Discussion but without comments returns 404
  - [x] In Discussion with comments returns profile page

### CODE REVIEW

**Review by coldlink — APPROVED** [review]
Good work and the best solution for now 👍

A question, but doesn't stop it going out:
How many extra calls do you think we'd get to discussion api as a result of this, and would there be an increase to the page load time as a result?
Since we're making an extra call on each profile page view, would be interesting to see what impact the DAPI call would have.

File: identity/app/controllers/PublicProfileController.scala [hunk 1]
```diff
@@ -13,39 +13,56 @@ import com.gu.identity.model.User
...
     activityType,
     identityApiClient.userFromVanityUrl(vanityUrl)
   )
 
-  def renderProfileFromId(id: String, activityType: String): Action[AnyContent] = renderPublicProfilePage("/user/id/"+id, activityType, identityApiClient.user(id))
+  def renderProfileFromId(id: String, activityType: String): Action[AnyContent] = renderPublicProfilePage("/user/id/" + id, activityType, identityApiClient.user(id))
 
   def renderPublicProfilePage(url: String, activityType: String, futureUser: => Future[Response[User]]): Action[AnyContent] = Action.async { implicit request =>
-    futureUser.map {
+    futureUser.flatMap {
```
Comment (dominickendrick) [inline identity/app/controllers/PublicProfileController.scala]: Could we add some comments here as to why this is working this way ?

File: identity/app/controllers/PublicProfileController.scala [hunk 2]
```diff
@@ -13,39 +13,56 @@ import com.gu.identity.model.User
...
+          case Left(error) =>
+            logger.info(s"public profile page returned error: ${error.message}")
+            NotFound(views.html.errors._404())
+
+          case Right(false) =>
+            logger.info(s"public profile page returned error: user ${user.id} found in Discussion but does not have a public profile")
+            NotFound(views.html.errors._404())
+
+          case Right(true) =>
+            val title = user.publicFields.username.fold("public profile")(username => s"$username's public profile") // TODO
```
Comment (dominickendrick) [inline identity/app/controllers/PublicProfileController.scala]: Is this still TODO ??

File: identity/app/controllers/PublicProfileController.scala [hunk 3]
```diff
@@ -13,39 +13,56 @@ import com.gu.identity.model.User
...
+        }
 
       case Right(user) =>
-        val title = user.publicFields.username.fold("public profile")(username => s"$username's public profile")
-        implicit val identityPage: IdentityPage = IdentityPage(url,  title, usesGuardianHeader = true)
-        Cached(60)(
-          RevalidatableResult.Ok(
-            IdentityHtmlPage.html(
-              views.html.publicProfilePage(identityPage, idRequestParser(request), idUrlBuilder, user, activityType)
+        discussionApi.userHasPublicProfile(user.id).value map {
```
Comment (dominickendrick) [inline identity/app/controllers/PublicProfileController.scala]: Can this be `value.map` just for convention

File: identity/app/services/DiscusssionApiService.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+case class DiscussionApiServiceException(message: String) extends Throwable
+
+class DiscussionApiService(wsClient: WSClient, config: conf.IdentityConfiguration)(implicit executionContext: ExecutionContext) extends SafeLogging {
+
+  private def GET(urlPath: String): Future[WSResponse] = {
+    wsClient
+      .url(config.discussionApiUrl + urlPath)
+      .withRequestTimeout(1.seconds)
```
Comment (dominickendrick) [inline identity/app/services/DiscusssionApiService.scala]: Is this the correct threshold for response from the API ? Could it be longer ?

File: identity/app/services/DiscusssionApiService.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import play.api.libs.json.{Format, JsValue, Json}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import play.api.libs.ws.{WSClient, WSResponse}
+import utils.SafeLogging
+import cats.implicits._
+import cats.data.EitherT
+
+// A user will exist in Discussion if they have commented or recommended a comment.
+// DAPI does not expose recommendation stats by user, so we only rely on comments to decide whether or not a user in Discussion should have a public profile
```
Comment (dominickendrick) [inline identity/app/services/DiscusssionApiService.scala]: Could we add a point that says `users need 1 or more comments to have a public profile page` so its explicit what the rule is?

**Review by dominickendrick — APPROVED** [review]
Looks good, apart from minor comments : )

File: identity/app/services/DiscusssionApiService.scala (hunk 4 above)
Comment (vlbee) [reply identity/app/services/DiscusssionApiService.scala]: I've updated this to 3sec. Latency spikes in Discussion metrics have hit up to 1sec in past 2 weeks.

**Comment by vlbee** [comment]
@coldlink  Frontend/identity would be making is making ~10k-17k/hour with peaks at about 500 reqs/min to Discussion

**Pull request merged**